src/controller/controller.c

```diff
@@ -444,6 +481,82 @@ static bool controller_setup_node_connection_handler(Controller *controller) {
         return true;
 }
 
+static int controller_reset_heartbeat_timer(Controller *controller, sd_event_source **event_source);
+
+static int controller_heartbeat_timer_callback(
+                sd_event_source *event_source, UNUSED uint64_t usec, void *userdata) {
+        Controller *controller = (Controller *) userdata;
+        Node *node = NULL;
+        int r = 0;
+
+        LIST_FOREACH(nodes, node, controller->nodes) {
+                uint64_t diff = 0;
+                uint64_t now = 0;
+
+                if (!node_is_online(node)) {
+                        continue;
+                }
+
+                now = get_time_micros();
+                if (now == 0) {
+                        return -EINVAL;
+                }
+
+                if (now < node->last_seen) {
+                        bc_log_error("Clock skew detected");
+                        return -EINVAL;
+                }
+
+                diff = now - node->last_seen;
+                if (diff > (uint64_t) controller->heartbeat_threshold_msec * USEC_PER_MSEC) {
+                        node_disconnect(node);
+                }
+        }
+
+        r = controller_reset_heartbeat_timer(controller, &event_source);
+        if (r < 0) {
+                bc_log_errorf("Failed to reset controller heartbeat timer: %s", strerror(-r));
+                return r;
+        }
+
+        return 0;
+}
+
+static int controller_reset_heartbeat_timer(Controller *controller, sd_event_source **event_source) {
+        return event_reset_time_relative(
+                        controller->event,
+                        event_source,
+                        CLOCK_BOOTTIME,
+                        controller->heartbeat_interval_msec * USEC_PER_MSEC,
+                        0,
+                        controller_heartbeat_timer_callback,
+                        controller,
+                        0,
+                        "controller-heartbeat-timer-source",
+                        false);
+}
+
+static int controller_setup_heartbeat_timer(Controller *controller) {
+        _cleanup_(sd_event_source_unrefp) sd_event_source *event_source = NULL;
+        int r = 0;
+
+        assert(controller);
+
+        if (controller->heartbeat_interval_msec <= 0) {
+                bc_log_warnf("Heartbeat disabled since interval is '%d', ",
+                             controller->heartbeat_interval_msec);
```

Comment:
```suggestion
                bc_log_warnf("Heartbeat disabled since configured interval '%d' is <=0",
                             controller->heartbeat_interval_msec);
```

Again, not related to the PR, but could you change this as well [for the agent here](https://github.com/eclipse-bluechi/bluechi/blob/main/src/agent/agent.c#L190).

Reply:
done